### PR TITLE
mgr/dashboard: empty grafana panels for performance of daemons 

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/61792
Signed-off-by: avanthakkar <avanjohn@gmail.com>

Before:
```
# TYPE ceph_osd_numpg gauge
 ceph_osd_numpg{ceph_daemon="ceph-osd.0"} 113
 ceph_osd_numpg{ceph_daemon="ceph-osd.1"} 113
 ceph_osd_numpg{ceph_daemon="ceph-osd.2"} 113
```
After: 
/metrics output(removed `ceph-` prefix from ceph_daemon label): 
```# HELP ceph_osd_numpg Placement groups
# TYPE ceph_osd_numpg gauge
ceph_osd_numpg{ceph_daemon="osd.0"} 113
ceph_osd_numpg{ceph_daemon="osd.1"} 113
ceph_osd_numpg{ceph_daemon="osd.2"} 113
```
Also the changes fixes the instance_id used for `rgw_metadata` metrics, when orch is available and matches with the one coming from ceph-exporter.

**Fixes the OSD grafana panels:**

Screenshot from 2023-07-01 15-04-31

Screenshot from 2023-07-01 15-03-08

**Fixes the RGW grafana panels:**
- RGW Overview: 
Screenshot from 2023-07-07 00-04-31

- RGW instance details:
Screenshot from 2023-07-07 00-05-22

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by Pegonzal** [comment]
jenkins test dashboard

File: src/exporter/DaemonMetricCollector.cc [hunk 1]
```diff
@@ -285,18 +288,37 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
   std::string new_metric_name;
   labels_t labels;
   new_metric_name = metric_name;
+  std::string ceph_daemon_prefix = "ceph-";
```
Comment (jmolmo) [inline src/exporter/DaemonMetricCollector.cc:291]: We have discover this with "ceph_osd_numpg" metrics. Are you sure this is applicable to all the rest of metrics?

**Review by jmolmo — CHANGES_REQUESTED** [review]
What we are seeing is that we have a regression because the "pattern" used in the label values is different from the previous one.
In order to avoid more regressions like this, it would be nice to have a "unit test" like the one used in the "promethize" function to verify that ceph exporter is producing the same kind of labels with the same values patterns we have using the prometheus manager module.

@avanthakkar do you think that is possible?

File: src/exporter/DaemonMetricCollector.cc (hunk 1 above)
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:291]: > We have discover this with "ceph_osd_numpg" metrics. Are you sure this is applicable to all the rest of metrics?

Yes. I've attached the screenshot as well. For fixing rgw related panels we need additional and major changes which I'm addressing here https://github.com/ceph/ceph/pull/52325/files. 
And for addressing the testing we already have a  tracker and as discussed before we agreed to address it in a separate PR.

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: src/exporter/DaemonMetricCollector.cc [hunk 2]
```diff
@@ -285,18 +285,34 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
...
+  std::string ceph_client_prefix = "client.";
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:289]:
```suggestion
  const std::string ceph_daemon_prefix = "ceph-";
  const std::string ceph_client_prefix = "client.";
```

File: src/exporter/DaemonMetricCollector.cc [hunk 3]
```diff
@@ -285,18 +285,34 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
...
+    std::string item;
+    while (std::getline(ss, item, '.')) {
+        elems.push_back(item);
+    }
+    labels["instance_id"] = quote(elems[3]);
+  } else {
+    labels.insert({"ceph_daemon", quote(daemon_name)});
   }
-  else if (daemon_name.find("rbd-mirror") != std::string::npos) {
+  if (daemon_name.find("rbd-mirror") != std::string::npos) {
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:315]: Why is it moved to another if chain?

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:315]: All ceph daemons requires `ceph_daemon` label except the rgw so I changed some conditions here.

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Review by pereman2 — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by cloudbehl** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by pereman2 — DISMISSED** [review]
[ RUN      ] Exporter.check_labels_and_metric_name
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/exporter/test_exporter.cc:684: Failure
Expected equality of these values:
  result.first
    Which is: { ("ceph_daemon", "\"osd.0\"") }
  labels_and_name_itr->first
    Which is: { ("ceph_daemon", "osd.0") }
[  FAILED  ] Exporter.check_labels_and_metric_name (0 ms)

**Comment by avanthakkar** [comment]
> 

Fixed

**Comment by avanthakkar** [comment]
jenkins test make check arm64

**Comment by avanthakkar** [comment]
> What we are seeing is that we have a regression because the "pattern" used in the label values is different from the previous one. In order to avoid more regressions like this, it would be nice to have a "unit test" like the one used in the "promethize" function to verify that ceph exporter is producing the same kind of labels with the same values patterns we have using the prometheus manager module.
> 
> @avanthakkar do you think that is possible?

Addedd a unit test for it.

File: src/exporter/DaemonMetricCollector.cc [hunk 4]
```diff
@@ -291,18 +291,34 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
...
-    labels["instance_id"] = quote("rgw." + tmp.substr(0, pos));
+    // fetch intance_id for e.g. "okbvtv" from daemon_name=rgw.foo.ceph-node-00.okbvtv 
+    std::vector<std::string> elems;
+    std::stringstream ss;
+    ss.str(daemon_name);
+    std::string item;
+    while (std::getline(ss, item, '.')) {
+        elems.push_back(item);
+    }
+    labels["instance_id"] = quote(elems[3]);
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:317]: I don't think a vector of strings is needed to only retrieve the last one of the values on a '.' delimitered list. You can just count the last value and use it. Maybe there could be a out of range error in elems[3].

File: src/test/exporter/test_exporter.cc [hunk 5]
```diff
@@ -662,3 +665,23 @@ TEST(Exporter, promethize) {
...
 }
 
+TEST(Exporter, check_labels_and_metric_name) {
+  std::vector<std::pair<std::string, std::string>> counters_data;
+  counters_data.emplace_back("ceph-osd.0", "ceph_osd_numpg");
+  counters_data.emplace_back("ceph-client.rgw.foo.ceph-node-00.okbvtv", "ceph_rgw_get");
+
+  std::vector<std::pair<labels_t, std::string>> labels_and_name;
+  labels_and_name.emplace_back(labels_t{{"ceph_daemon", "\"osd.0\""}}, "ceph_osd_numpg");
+  labels_and_name.emplace_back(labels_t{{"instance_id", "\"okbvtv\""}}, "ceph_rgw_get");
```
Comment (pereman2) [inline src/test/exporter/test_exporter.cc:675]: This could be static.

**Review by jmolmo — APPROVED** [review]
LGTM. (Just address the changes requested by Pere)

File: src/exporter/DaemonMetricCollector.cc (hunk 4 above)
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:317]: Ah yes! Thanks for noticing! it makes things much simpler, will update it.

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc]: Done!

**Review by pereman2 — APPROVED** [review]

File: src/exporter/DaemonMetricCollector.cc (hunk 4 above)
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc]: We actually need to fetch the 4th element only as this fails if daemon name is like `ceph-client.rgw.foo.ceph-node-00.hrgsea.2.94739968030880`

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc]: I've opened a PR to bring back these changes.

**Pull request merged**